drivers/xen/regions.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_arch_interface.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/arch_interface.h>
+#include <zephyr/sys/mem_blocks.h>
+#include <zephyr/xen/generic.h>
+
+LOG_MODULE_REGISTER(xen_regions);
+
+#define EXTENDED_REGIONS_IDX 1
+#define XEN_HYP_NODE DT_INST(0, xen_xen)
+
+static K_MUTEX_DEFINE(ext_regions_lock);
+
+struct extended_regions {
+	sys_mem_blocks_t *block;
+	uint64_t phys_addr;
+	uint64_t size;
+};
+
+/*
+ * At least one extended region should be provided.
+ * If there is no extended regions - alternative allocation should be implemented
+ * by the driver user and driver should be disabled.
+ */
+BUILD_ASSERT(DT_NUM_REGS(XEN_HYP_NODE) > EXTENDED_REGIONS_IDX,
+	     "No Extended regions were provided in DT");
+
+#define NOT_OVERLAPPING_BOUNDARIES(x_start, x_end, x_size, y_start, y_end, y_size) \
```

Comment:
(note) there is more simple solution, i.e. less conditions:
`(x1 - y2) * (y1 - x2) > 0`
or if you dislike multiplication you can use this one but with one extra comparison:
`(a2 > b1) && (a1 < b2)`

Reply:
The first variant will require all operands to be signed. The second - Are you sure it covers all cases? 
If yes - ok I will add it.

Reply:
if you want include toches of ranges just change `</>` to `<=/>=`
![Screenshot from 2023-05-26 19-06-19](https://github.com/xen-troops/zephyr/assets/27217224/7ef0b60d-dc2b-4101-83bf-077463e526b0)

Reply:
in variant 2 what if a2 = b1? it will return false.

Reply:
false mean there isn't intersection,
case a2 == b1 it is touch one range to another if you want to include touches as intersections, as I wrote in prev comment
just change conditions from `>` to `>=` and `<` to `<=`

Reply:
Sorry. I can't check each approach you suggesting not because code is wrong, but because to like your approach more. If you wan't me to use your approach please give me the final variant.